File: artifacts/api-server/src/routes/analytics.py
"""Analytics routes — grade distribution, at-risk, performance, trends, completion."""
from typing import Optional
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from src.db.session import get_db
from src.db.models import GradeRow, AssignmentRow, CourseRow, EnrollmentRow, StudentRow
from src.domain.grade_utils import to_percent, score_to_letter, letter_to_points, risk_level

router = APIRouter()


@router.get("/analytics/grade-distribution")
def grade_distribution(course_id: int = Query(...), db: Session = Depends(get_db)):
    data = (
        db.query(GradeRow.score, AssignmentRow.max_score)
        .join(AssignmentRow, GradeRow.assignment_id == AssignmentRow.id)
        .filter(AssignmentRow.course_id == course_id)
        .all()
    )
    pcts = [to_percent(float(s), float(m)) for s, m in data]
    buckets = [
        {"range": f"{lo}-{hi - 1}", "min": lo, "max": hi}
        for lo, hi in [(i * 10, i * 10 + 10) for i in range(10)]
    ]
    buckets[-1]["range"] = "90-100"
    buckets[-1]["max"] = 101
    result_buckets = []
    for b in buckets:
        c = sum(1 for p in pcts if b["min"] <= p < b["max"])
        result_buckets.append({
            "range": b["range"],
            "count": c,
            "percentage": round((c / len(pcts)) * 1000) / 10 if pcts else 0,
        })
    letter_map = {"A": 0, "B": 0, "C": 0, "D": 0, "F": 0}
    for p in pcts:
        letter_map[score_to_letter(p)] += 1
    letter_counts = [
        {
            "letter": l,
            "count": c,
            "percentage": round((c / len(pcts)) * 1000) / 10 if pcts else 0,
        }
        for l, c in letter_map.items()
    ]
    return {"course_id": course_id, "buckets": result_buckets, "letter_counts": letter_counts}


@router.get("/analytics/at-risk")
def at_risk_students(
    course_id: Optional[int] = Query(None),
    threshold: Optional[float] = Query(None),
    db: Session = Depends(get_db),
):
    min_pct = threshold if threshold is not None else 70.0
    courses = (
        db.query(CourseRow).filter(CourseRow.id == course_id).all()
        if course_id
        else db.query(CourseRow).all()
    )
    result = []
    for course in courses:
        enrollments = db.query(EnrollmentRow).filter(EnrollmentRow.course_id == course.id).all()
        assignments = db.query(AssignmentRow).filter(AssignmentRow.course_id == course.id).all()
        for enr in enrollments:
            student = db.query(StudentRow).filter(StudentRow.id == enr.student_id).first()
            if not student:
                continue
            weighted_score = 0.0
            total_weight = 0.0
            missing = 0
            for asgn in assignments:
                grade = db.query(GradeRow).filter(
                    GradeRow.student_id == enr.student_id,
                    GradeRow.assignment_id == asgn.id,
                ).first()
                if grade:
                    pct = to_percent(float(grade.score), float(asgn.max_score))
                    weighted_score += pct * float(asgn.weight)
                    total_weight += float(asgn.weight)
                else:
                    missing += 1
            if total_weight > 0:
                current_pct = weighted_score / total_weight
                if current_pct < min_pct:
                    result.append({
                        "student_id": student.id,
                        "student_name": student.name,
                        "current_grade": round(current_pct * 10) / 10,
                        "letter_grade": score_to_letter(current_pct),
                        "course_name": course.name,
                        "course_id": course.id,
                        "assignments_missing": missing,
                        "risk_level": risk_level(current_pct),
                    })
            elif assignments:
                result.append({
                    "student_id": student.id,
                    "student_name": student.name,
                    "current_grade": 0.0,
                    "letter_grade": "F",
                    "course_name": course.name,
                    "course_id": course.id,
                    "assignments_missing": len(assignments),
                    "risk_level": "high",
                })
    return result


@router.get("/analytics/course-performance")
def course_performance(db: Session = Depends(get_db)):
    courses = db.query(CourseRow).order_by(CourseRow.name).all()
    result = []
    for c in courses:
        enr_count = db.query(EnrollmentRow).filter(EnrollmentRow.course_id == c.id).count()
        data = (
            db.query(GradeRow.score, AssignmentRow.max_score)
            .join(AssignmentRow, GradeRow.assignment_id == AssignmentRow.id)
            .filter(AssignmentRow.course_id == c.id)
            .all()
        )
        pcts = [to_percent(float(s), float(m)) for s, m in data]
        avg = sum(pcts) / len(pcts) if pcts else 0
        passing = sum(1 for p in pcts if p >= 60)
        pass_rate = (passing / len(pcts)) * 100 if pcts else 0
        result.append({
            "course_id": c.id,
            "course_name": c.name,
            "average_grade": round(avg * 10) / 10,
            "pass_rate": round(pass_rate * 10) / 10,
            "student_count": enr_count,
            "semester": c.semester,
            "difficulty_score": round((100 - avg) * 10) / 10 if avg > 0 else None,
        })
    return result


@router.get("/analytics/semester-trends")
def semester_trends(student_id: Optional[int] = Query(None), db: Session = Depends(get_db)):
    enrollments = (
        db.query(EnrollmentRow).filter(EnrollmentRow.student_id == student_id).all()
        if student_id
        else db.query(EnrollmentRow).all()
    )
    semester_map: dict = {}
    for enr in enrollments:
        course = db.query(CourseRow).filter(CourseRow.id == enr.course_id).first()
        if not course:
            continue
        assignments = db.query(AssignmentRow).filter(AssignmentRow.course_id == enr.course_id).all()
        weighted_score = 0.0
        total_weight = 0.0
        for asgn in assignments:
            q = db.query(GradeRow).filter(GradeRow.assignment_id == asgn.id)
            if student_id:
                q = q.filter(GradeRow.student_id == student_id)
            grades = q.all()
            if grades:
                avg_pct = sum(to_percent(float(g.score), float(asgn.max_score)) for g in grades) / len(grades)
                weighted_score += avg_pct * float(asgn.weight)
                total_weight += float(asgn.weight)
        if total_weight > 0:
            final_pct = weighted_score / total_weight
            gp = letter_to_points(score_to_letter(final_pct))
            existing = semester_map.get(enr.semester, {"totalPoints": 0.0, "totalCredits": 0, "courseCount": 0})
            semester_map[enr.semester] = {
                "totalPoints": existing["totalPoints"] + gp * course.credits,
                "totalCredits": existing["totalCredits"] + course.credits,
                "courseCount": existing["courseCount"] + 1,
            }
    return sorted(
        [
            {
                "semester": sem,
                "gpa": round((d["totalPoints"] / d["totalCredits"]) * 100) / 100 if d["totalCredits"] > 0 else 0,
                "courses_taken": d["courseCount"],
                "student_id": student_id,
            }
            for sem, d in semester_map.items()
        ],
        key=lambda x: x["semester"],
    )


@router.get("/analytics/assignment-completion")
def assignment_completion(course_id: Optional[int] = Query(None), db: Session = Depends(get_db)):
    q = db.query(AssignmentRow)
    if course_id is not None:
        q = q.filter(AssignmentRow.course_id == course_id)
    assignments = q.all()
    result = []
    for asgn in assignments:
        total = db.query(EnrollmentRow).filter(EnrollmentRow.course_id == asgn.course_id).count()
        submitted = db.query(GradeRow).filter(GradeRow.assignment_id == asgn.id).count()
        completion_rate = round((submitted / total) * 1000) / 10 if total > 0 else 0
        scores = db.query(GradeRow.score).filter(GradeRow.assignment_id == asgn.id).all()
        avg_score = None
        if scores:
            pcts = [to_percent(float(s[0]), float(asgn.max_score)) for s in scores]
            avg_score = round(sum(pcts) / len(pcts) * 10) / 10
        result.append({
            "assignment_id": asgn.id,
            "assignment_name": asgn.name,
            "type": asgn.type,
            "submitted_count": submitted,
            "total_enrolled": total,
            "completion_rate": completion_rate,
            "average_score": avg_score,
        })
    return result

File: artifacts/server/src/routes/analytics.ts
import { Router, type IRouter } from "express";
import { eq, and, count } from "drizzle-orm";
import {
  db, gradesTable, assignmentsTable, coursesTable, enrollmentsTable, studentsTable,
} from "@workspace/db";
import {
  GetGradeDistributionQueryParams,
  GetAtRiskStudentsQueryParams,
  GetSemesterTrendsQueryParams,
  GetAssignmentCompletionQueryParams,
} from "@workspace/api-zod";
import { scoreToLetter, toPercent, letterToPoints, riskLevel } from "../lib/gradeUtils";

const router: IRouter = Router();

// GET /analytics/grade-distribution?course_id=X
router.get("/analytics/grade-distribution", async (req, res): Promise<void> => {
  const query = GetGradeDistributionQueryParams.safeParse(req.query);
  if (!query.success) {
    res.status(400).json({ error: query.error.message });
    return;
  }

  const { course_id } = query.data as { course_id: number };

  const gradeData = await db
    .select({ score: gradesTable.score, maxScore: assignmentsTable.maxScore })
    .from(gradesTable)
    .innerJoin(assignmentsTable, eq(gradesTable.assignmentId, assignmentsTable.id))
    .where(eq(assignmentsTable.courseId, course_id));

  const pcts = gradeData.map((g) => toPercent(Number(g.score), Number(g.maxScore)));

  // Build 10-point buckets
  const bucketDefs = [
    { range: "0-9", min: 0, max: 10 },
    { range: "10-19", min: 10, max: 20 },
    { range: "20-29", min: 20, max: 30 },
    { range: "30-39", min: 30, max: 40 },
    { range: "40-49", min: 40, max: 50 },
    { range: "50-59", min: 50, max: 60 },
    { range: "60-69", min: 60, max: 70 },
    { range: "70-79", min: 70, max: 80 },
    { range: "80-89", min: 80, max: 90 },
    { range: "90-100", min: 90, max: 101 },
  ];

  const buckets = bucketDefs.map((b) => {
    const c = pcts.filter((p) => p >= b.min && p < b.max).length;
    return { range: b.range, count: c, percentage: pcts.length > 0 ? Math.round((c / pcts.length) * 1000) / 10 : 0 };
  });

  // Letter grade counts
  const letterMap: Record<string, number> = { A: 0, B: 0, C: 0, D: 0, F: 0 };
  pcts.forEach((p) => { letterMap[scoreToLetter(p)]++; });
  const letterCounts = Object.entries(letterMap).map(([letter, c]) => ({
    letter,
    count: c,
    percentage: pcts.length > 0 ? Math.round((c / pcts.length) * 1000) / 10 : 0,
  }));

  res.json({ course_id, buckets, letter_counts: letterCounts });
});

// GET /analytics/at-risk?course_id=X&threshold=Y
router.get("/analytics/at-risk", async (req, res): Promise<void> => {
  const query = GetAtRiskStudentsQueryParams.safeParse(req.query);
  if (!query.success) {
    res.status(400).json({ error: query.error.message });
    return;
  }

  const { course_id, threshold } = query.data as { course_id?: number; threshold?: number };
  const minPct = threshold ?? 70; // default: below 70% is at risk

  const courses = course_id
    ? await db.select().from(coursesTable).where(eq(coursesTable.id, course_id))
    : await db.select().from(coursesTable);

  const result = [];

  for (const course of courses) {
    const enrollments = await db
      .select()
      .from(enrollmentsTable)
      .where(eq(enrollmentsTable.courseId, course.id));

    for (const enr of enrollments) {
      const student = await db.select().from(studentsTable).where(eq(studentsTable.id, enr.studentId)).limit(1);
      if (!student[0]) continue;

      const assignments = await db.select().from(assignmentsTable).where(eq(assignmentsTable.courseId, course.id));
      let weightedScore = 0;
      let totalWeight = 0;
      let missing = 0;

      for (const asgn of assignments) {
        const grade = await db.select().from(gradesTable).where(and(eq(gradesTable.studentId, enr.studentId), eq(gradesTable.assignmentId, asgn.id))).limit(1);
        if (grade[0]) {
          const pct = toPercent(Number(grade[0].score), Number(asgn.maxScore));
          weightedScore += pct * Number(asgn.weight);
          totalWeight += Number(asgn.weight);
        } else {
          missing++;
        }
      }

      if (totalWeight > 0) {
        const currentPct = weightedScore / totalWeight;
        if (currentPct < minPct) {
          const letter = scoreToLetter(currentPct);
          result.push({
            student_id: student[0].id,
            student_name: student[0].name,
            current_grade: Math.round(currentPct * 10) / 10,
            letter_grade: letter,
            course_name: course.name,
            course_id: course.id,
            assignments_missing: missing,
            risk_level: riskLevel(currentPct),
          });
        }
      } else if (assignments.length > 0) {
        // Has assignments but no grades at all — high risk
        result.push({
          student_id: student[0].id,
          student_name: student[0].name,
          current_grade: 0,
          letter_grade: "F",
          course_name: course.name,
          course_id: course.id,
          assignments_missing: assignments.length,
          risk_level: "high" as const,
        });
      }
    }
  }

  res.json(result);
});

// GET /analytics/course-performance
router.get("/analytics/course-performance", async (req, res): Promise<void> => {
  const courses = await db.select().from(coursesTable).orderBy(coursesTable.name);

  const result = await Promise.all(
    courses.map(async (c) => {
      const enrollmentCount = await db.select({ cnt: count() }).from(enrollmentsTable).where(eq(enrollmentsTable.courseId, c.id));
      const gradeData = await db
        .select({ score: gradesTable.score, maxScore: assignmentsTable.maxScore })
        .from(gradesTable)
        .innerJoin(assignmentsTable, eq(gradesTable.assignmentId, assignmentsTable.id))
        .where(eq(assignmentsTable.courseId, c.id));

      const pcts = gradeData.map((g) => toPercent(Number(g.score), Number(g.maxScore)));
      const avg = pcts.length > 0 ? pcts.reduce((a, b) => a + b, 0) / pcts.length : 0;
      const passing = pcts.filter((p) => p >= 60).length;
      const passRate = pcts.length > 0 ? (passing / pcts.length) * 100 : 0;

      return {
        course_id: c.id,
        course_name: c.name,
        average_grade: Math.round(avg * 10) / 10,
        pass_rate: Math.round(passRate * 10) / 10,
        student_count: Number(enrollmentCount[0]?.cnt ?? 0),
        semester: c.semester,
        difficulty_score: avg > 0 ? Math.round((100 - avg) * 10) / 10 : null,
      };
    }),
  );

  res.json(result);
});

// GET /analytics/semester-trends?student_id=X
router.get("/analytics/semester-trends", async (req, res): Promise<void> => {
  const query = GetSemesterTrendsQueryParams.safeParse(req.query);
  if (!query.success) {
    res.status(400).json({ error: query.error.message });
    return;
  }

  const { student_id } = query.data as { student_id?: number };

  // Group enrollments by semester
  const enrollments = student_id
    ? await db.select().from(enrollmentsTable).where(eq(enrollmentsTable.studentId, student_id))
    : await db.select().from(enrollmentsTable);

  // Build map: semester -> {totalPoints, totalCredits}
  const semesterMap: Map<string, { totalPoints: number; totalCredits: number; courseCount: number }> = new Map();

  for (const enr of enrollments) {
    const course = await db.select().from(coursesTable).where(eq(coursesTable.id, enr.courseId)).limit(1);
    if (!course[0]) continue;

    const assignments = await db.select().from(assignmentsTable).where(eq(assignmentsTable.courseId, enr.courseId));
    let weightedScore = 0;
    let totalWeight = 0;

    for (const asgn of assignments) {
      const studentFilter = student_id ? and(eq(gradesTable.studentId, student_id), eq(gradesTable.assignmentId, asgn.id)) : eq(gradesTable.assignmentId, asgn.id);
      const grades = await db.select().from(gradesTable).where(studentFilter);
      if (grades.length > 0) {
        const avgPct = grades.reduce((sum, g) => sum + toPercent(Number(g.score), Number(asgn.maxScore)), 0) / grades.length;
        weightedScore += avgPct * Number(asgn.weight);
        totalWeight += Number(asgn.weight);
      }
    }

    if (totalWeight > 0) {
      const finalPct = weightedScore / totalWeight;
      const letter = scoreToLetter(finalPct);
      const gp = letterToPoints(letter);
      const existing = semesterMap.get(enr.semester) ?? { totalPoints: 0, totalCredits: 0, courseCount: 0 };
      semesterMap.set(enr.semester, {
        totalPoints: existing.totalPoints + gp * course[0].credits,
        totalCredits: existing.totalCredits + course[0].credits,
        courseCount: existing.courseCount + 1,
      });
    }
  }

  const result = Array.from(semesterMap.entries())
    .map(([semester, data]) => ({
      semester,
      gpa: data.totalCredits > 0 ? Math.round((data.totalPoints / data.totalCredits) * 100) / 100 : 0,
      courses_taken: data.courseCount,
      student_id: student_id ?? null,
    }))
    .sort((a, b) => a.semester.localeCompare(b.semester));

  res.json(result);
});

// GET /analytics/assignment-completion?course_id=X
router.get("/analytics/assignment-completion", async (req, res): Promise<void> => {
  const query = GetAssignmentCompletionQueryParams.safeParse(req.query);
  if (!query.success) {
    res.status(400).json({ error: query.error.message });
    return;
  }

  const { course_id } = query.data as { course_id?: number };

  const assignments = course_id
    ? await db.select().from(assignmentsTable).where(eq(assignmentsTable.courseId, course_id))
    : await db.select().from(assignmentsTable);

  const result = await Promise.all(
    assignments.map(async (asgn) => {
      const enrollmentCount = await db.select({ cnt: count() }).from(enrollmentsTable).where(eq(enrollmentsTable.courseId, asgn.courseId));
      const total = Number(enrollmentCount[0]?.cnt ?? 0);
      const submitted = await db.select({ cnt: count() }).from(gradesTable).where(eq(gradesTable.assignmentId, asgn.id));
      const submittedCount = Number(submitted[0]?.cnt ?? 0);
      const completionRate = total > 0 ? Math.round((submittedCount / total) * 1000) / 10 : 0;

      const gradeData = await db.select({ score: gradesTable.score }).from(gradesTable).where(eq(gradesTable.assignmentId, asgn.id));
      let avgScore: number | null = null;
      if (gradeData.length > 0) {
        const pcts = gradeData.map((g) => toPercent(Number(g.score), Number(asgn.maxScore)));
        avgScore = Math.round((pcts.reduce((a, b) => a + b, 0) / pcts.length) * 10) / 10;
      }

      return {
        assignment_id: asgn.id,
        assignment_name: asgn.name,
        type: asgn.type,
        submitted_count: submittedCount,
        total_enrolled: total,
        completion_rate: completionRate,
        average_score: avgScore,
      };
    }),
  );

  res.json(result);
});

export default router;
